File: Data Structures & Algorithms/three-integer-sum/submission-0.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        pairs = []
        nums = sorted(nums)
        for i, num in enumerate(nums):
            left = i+1
            right = len(nums) - 1
            rem = 0 - num
            while(left < right):
                if (nums[left] + nums[right] > rem):
                    right -= 1
                elif (nums[left] + nums[right] < rem):
                    left += 1
                else:
                    pairs.append(sorted([nums[i], nums[left], nums[right]]))
                    left += 1
                    right -= 1

        unique_lists = [list(x) for x in set(tuple(l) for l in pairs)]
        return unique_lists

File: Data Structures & Algorithms/search-2d-matrix/submission-0.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int n = matrix.size(), m = matrix[0].size();
        int l = 0, r = n-1;
        while(l <= r) {
            int mid = l + (r - l)/2;
            if(target >= matrix[mid][0] && target <= matrix[mid][m-1] ) {
                int left = 0, right = m - 1;
                while(left <= right) {
                    int m = left + (right-left)/ 2;
                    if(matrix[mid][m] == target) return true;
                    else if(matrix[mid][m] > target) right = m-1;
                    else left = m+1;
                }
                return false;
            }
            else if(target > matrix[mid][m-1]) l = mid+1;
            else r = mid - 1;
        }
        return false;
    }
};
